### PR TITLE
Discard sk if ivk == 0

### PR DESCRIPTION
**Pull request by arcalinea**
Addresses #3319

### CODE REVIEW

**Comment by ebfull** [comment]
Cool! ACK

File: src/zcash/Address.cpp [hunk 1]
```diff
@@ -67,8 +67,23 @@ SaplingIncomingViewingKey SaplingFullViewingKey::in_viewing_key() const {
...
 }
 
+bool SaplingFullViewingKey::is_valid() const {
+    uint256 ivk;
+    librustzcash_crh_ivk(ak.begin(), nk.begin(), ivk.begin());
+    if (!ivk.IsNull()) {
+        return true;
+    } else {
+        return false;
+    }
```
Comment (str4d) [inline src/zcash/Address.cpp]: Could simplify this to `return !ivk.IsNull();`

File: src/zcash/Address.cpp [hunk 2]
```diff
@@ -67,8 +67,23 @@ SaplingIncomingViewingKey SaplingFullViewingKey::in_viewing_key() const {
...
+}
+
 SaplingSpendingKey SaplingSpendingKey::random() {
-    return SaplingSpendingKey(random_uint256());
+    while (true) {
+        auto sk = SaplingSpendingKey(random_uint256());
+        if(sk.full_viewing_key().is_valid()) {
```
Comment (daira) [inline src/zcash/Address.cpp]: Nit: space after ``if``.

**Comment by arcalinea** [comment]
Addressed changes

**Comment by str4d** [comment]
@zkbot r+

**Pull request merged**